src/packaging/metadata.py

```diff
@@ -406,3 +427,193 @@ def parse_email(data: Union[bytes, str]) -> Tuple[RawMetadata, Dict[str, List[st
+        return value
+
+    def _process_metadata_version(self, value: str) -> _MetadataVersion:
+        if value not in {"1.0", "1.1", "1.2", "2.0", "2.1", "2.2", "2.3"}:
+            raise self._invalid_metadata(f"{value!r} is not a valid metadata version")
+        return cast(_MetadataVersion, value)
+
+    def _process_description_content_type(self, value: str) -> str:
+        content_types = {"text/plain", "text/x-rst", "text/markdown"}
+        message = email.message.EmailMessage()
+        message["content-type"] = value
+        content_type, parameters = (
+            message.get_content_type(),  # Defaults to `text/plain` if not parseable.
+            message["content-type"].params,
+        )
+        # Check if content-type is valid or defaulted to `text/plain` and thus was
+        # not parseable.
+        if content_type not in content_types or content_type not in value:
+            raise self._invalid_metadata(
+                f"{{field}} must be one of {list(content_types)}, not {value!r}"
+            )
+
+        charset = parameters.get("charset", "UTF-8")
+        if charset != "UTF-8":
+            raise self._invalid_metadata(
+                f"{{field}} can only specify the UTF-8 charset, not {list(charset)}"
+            )
+
+        markdown_variants = {"GFM", "CommonMark"}
+        variant = parameters.get("variant", "GFM")  # Use an acceptable default.
+        if content_type == "text/markdown" and variant not in markdown_variants:
+            raise self._invalid_metadata(
+                f"valid Markdown variants for {{field}} are {list(markdown_variants)}, "
+                f"not {variant!r}",
+            )
+        return value
+
+    def _process_dynamic(self, value: List[str]) -> List[str]:
+        for dynamic_field in map(str.lower, value):
+            if dynamic_field in {"name", "version", "metadata-version"}:
+                raise self._invalid_metadata(
+                    f"{value!r} is not allowed as a dynamic field"
+                )
+            elif dynamic_field not in _EMAIL_TO_RAW_MAPPING:
+                raise self._invalid_metadata(f"{value!r} is not a valid dynamic field")
+        return list(map(str.lower, value))
+
+    @staticmethod
+    def _process_provides_extra(
+        value: List[str],
+    ) -> List[utils.NormalizedName]:
+        return [utils.canonicalize_name(name, validate=True) for name in value]
+
+    _process_requires_python = staticmethod(specifiers.SpecifierSet)
+
+    @staticmethod
+    def _process_requires_dist(
+        value: List[str],
+    ) -> List[requirements.Requirement]:
+        return list(map(requirements.Requirement, value))
+
+
+class Metadata:
+    _raw: RawMetadata
+
+    @classmethod
+    def from_raw(cls, data: RawMetadata) -> "Metadata":
+        ins = cls()
+        ins._raw = data
+        return ins
+
+    @classmethod
+    def from_email(cls, data: Union[bytes, str]) -> "Metadata":
```

Comment:
While `from_email` is obvious for those who know about how `PKG-INFO`/`METADATA` files are constructed (email headers), as an end-user I would expect something like `from_metadata_file(file)` or something similar. However, I can understand if this could be too much. Basically, I would expect `packaging.metadata` to be a semi-drop-in replacement for [`pkginfo`](https://pythonhosted.org/pkginfo/)

Reply:
I'm not totally opposed to a nicer name, but it matches https://packaging.pypa.io/en/stable/metadata.html#packaging.metadata.parse_email currently.

As for the the `file` argument, we are specifically taking a [sans-I/O](https://sans-io.readthedocs.io/) approach to this API (and this project overall), so we won't have an API that takes a file path (closest would be for a file-like object, but even then I don't know if it's worth it since people could be getting this info via an HTTP request to PyPI).

Reply:
I understand the sans-I/O part, it makes total sense. I would still go for a nicer name though. However, I have no idea for a good one. `from_metadata_content(content: str)`? Doesn't sound right. 

On the other hand, most users would probably be knowledgeable enough to understand the `email` part of the name. Especially if it matches the current API.